File: leetcode/solutions/0594.LongestHarmoniousSubsequence/findLHS.ts
export default function findLHS(nums: number[]): number {
  const valueCountMap = new Map<number, number>();

  for (const num of nums.values()) {
    valueCountMap.set(num, (valueCountMap.get(num) ?? 0) + 1);
  }

  let max = 0;

  for (const [num, count] of valueCountMap) {
    const nextCount = valueCountMap.get(num + 1);

    if (nextCount != null) {
      max = Math.max(max, count + nextCount);
    }
  }

  return max;
}

File: leetcode/solutions/0594.LongestHarmoniousSubsequence/findLHS.go
package solution

import (
	"math"
)

func findLHS(nums []int) int {
	valueCountMap := make(map[int]int)

	for _, num := range nums {
		valueCountMap[num]++
	}

	max := 0.0

	for i, num := range valueCountMap {
		if next, ok := valueCountMap[i+1]; ok {
			max = math.Max(max, float64(num + next))
		}
	}

	return int(max)
}
